File: MiniDOMOffice/Magazzino/Articolo/ArticoloXListino/ArticoloXListino.vb
Imports minidom.Databases
Imports minidom.Sistema
Imports minidom.Anagrafica



Partial Public Class Office

    ''' <summary>
    ''' Definisce il tipo prezzo
    ''' </summary>
    ''' <remarks></remarks>
    Public Enum TipoPrezzoListino As Integer
        FISSO = 0
    End Enum

    ''' <summary>
    ''' Definisce la relazione tra un articolo ed un listino prezzi
    ''' </summary>
    ''' <remarks></remarks>
    <Serializable> _
    Public Class ArticoloXListino
        Inherits DBObject

        Private m_TipoPrezzo As TipoPrezzoListino
        Private m_IDArticolo As Integer
        <NonSerialized> Private m_Articolo As Articolo
        Private m_IDListino As Integer
        <NonSerialized> Private m_Listino As Listino
        Private m_Flags As Integer

        Public Sub New()
            Me.m_IDArticolo = 0
            Me.m_Articolo = Nothing
            Me.m_IDListino = 0
            Me.m_Listino = Nothing
            Me.m_Flags = 0
            Me.m_TipoPrezzo = TipoPrezzoListino.FISSO
        End Sub

        ''' <summary>
        ''' Restituisce o imposta il tipo prezzo
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public Property TipoPrezzo As TipoPrezzoListino
            Get
                Return Me.m_TipoPrezzo
            End Get
            Set(value As TipoPrezzoListino)
                Dim oldValue As TipoPrezzoListino = Me.m_TipoPrezzo
                If (oldValue = value) Then Exit Property
                Me.m_TipoPrezzo = value
                Me.DoChanged("TipoPrezzo", value, oldValue)
            End Set
        End Property

        ''' <summary>
        ''' Restituisce o imposta l'articolo a cui è associato il codice
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public Property Articolo As Articolo
            Get
                SyncLock Me
                    If (Me.m_Articolo Is Nothing) Then Me.m_Articolo = Office.Articoli.GetItemById(Me.m_IDArticolo)
                    Return Me.m_Articolo
                End SyncLock
            End Get
            Set(value As Articolo)
                Dim oldValue As Articolo
                SyncLock Me
                    oldValue = Me.m_Articolo
                    If (oldValue Is value) Then Exit Property
                    Me.m_Articolo = value
                    Me.m_IDArticolo = GetID(value)
                End SyncLock
                Me.DoChanged("Articolo", value, oldValue)
            End Set
        End Property

        Public Property IDArticolo As Integer
            Get
                Return GetID(Me.m_Articolo, Me.m_IDArticolo)
            End Get
            Set(value As Integer)
                Dim oldValue As Integer = Me.IDArticolo
                If (oldValue = value) Then Exit Property
                Me.m_IDArticolo = 0
                Me.m_Articolo = Nothing
                Me.DoChanged("IDArticolo", value, oldValue)
            End Set
        End Property

        Protected Friend Overridable Sub SetArticolo(ByVal value As Articolo)
            Me.m_Articolo = value
            Me.m_IDArticolo = GetID(value)
        End Sub

        ''' <summary>
        ''' Restituisce o imposta il magazzino
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public Property Listino As Listino
            Get
                SyncLock Me
                    If Me.m_Listino Is Nothing Then Me.m_Listino = Office.Listini.GetItemById(Me.m_IDListino)
                    Return Me.m_Listino
                End SyncLock
            End Get
            Set(value As Listino)
                Dim oldValue As Listino
                SyncLock Me
                    oldValue = Me.m_Listino
                    If (oldValue Is value) Then Exit Property
                    Me.m_Listino = value
                    Me.m_IDListino = GetID(value)
                End SyncLock
                Me.DoChanged("Listino", value, oldValue)
            End Set
        End Property

        Public Property IDListino As Integer
            Get
                Return GetID(Me.m_Listino, Me.m_IDListino)
            End Get
            Set(value As Integer)
                Dim oldValue As Integer = Me.IDListino
                If (oldValue = value) Then Exit Property
                Me.m_IDListino = value
                Me.m_Listino = Nothing
                Me.DoChanged("IDListino", value, oldValue)
            End Set
        End Property
         
        ''' <summary>
        ''' Restituisce o imposta una serie di flags associati al codice articolo
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public Property Flags As Integer
            Get
                Return Me.m_Flags
            End Get
            Set(value As Integer)
                Dim oldValue As Integer = Me.m_Flags
                If (oldValue = value) Then Exit Property
                Me.m_Flags = value
                Me.DoChanged("Flags", value, oldValue)
            End Set
        End Property

        Public Function CalcolaPrezzoVendita(ByVal docuemnto As Object) As Decimal
            Return 0
        End Function


        Protected Overrides Sub XMLSerialize(writer As XML.XMLWriter)
            writer.WriteAttribute("IDArticolo", Me.IDArticolo)
            writer.WriteAttribute("IDListino", Me.IDListino)
            writer.WriteAttribute("Flags", Me.m_Flags)
            writer.WriteAttribute("TipoPrezzo", Me.m_TipoPrezzo)
            MyBase.XMLSerialize(writer)
        End Sub

        Protected Overrides Sub SetFieldInternal(fieldName As String, fieldValue As Object)
            Select Case fieldName
                Case "IDArticolo" : Me.m_IDArticolo = XML.Utils.Serializer.DeserializeInteger(fieldValue)
                Case "IDListino" : Me.m_IDListino = XML.Utils.Serializer.DeserializeInteger(fieldValue)
                Case "Flags" : Me.m_Flags = XML.Utils.Serializer.DeserializeInteger(fieldValue)
                Case "TipoPrezzo" : Me.m_TipoPrezzo = XML.Utils.Serializer.DeserializeInteger(fieldValue)
                Case Else : MyBase.SetFieldInternal(fieldName, fieldValue)
            End Select
        End Sub


        Public Overrides Function GetTableName() As String
            Return "tbl_OfficeArticoliListino"
        End Function

        Protected Overrides Function GetConnection() As CDBConnection
            Return Office.Database
        End Function

        Public Overrides Function GetModule() As CModule
            Return Nothing
        End Function

        Protected Overrides Function LoadFromRecordset(reader As DBReader) As Boolean
            Me.m_IDArticolo = reader.Read("IDArticolo", Me.m_IDArticolo)
            Me.m_IDListino = reader.Read("IDListino", Me.m_IDListino)
            Me.m_Flags = reader.Read("Flags", Me.m_Flags)
            Me.m_TipoPrezzo = reader.Read("TipoPrezzo", Me.m_TipoPrezzo)
            Return MyBase.LoadFromRecordset(reader)
        End Function

        Protected Overrides Function SaveToRecordset(writer As DBWriter) As Boolean
            writer.Write("IDArticolo", Me.IDArticolo)
            writer.Write("IDListino", Me.IDListino)
            writer.Write("Flags", Me.m_Flags)
            writer.Write("TipoPrezzo", Me.m_TipoPrezzo)
            Return MyBase.SaveToRecordset(writer)
        End Function

    
    End Class

End Class

File: MiniDOMOffice/Magazzino/Articolo/ArticoloXListino/ArticoloXListino.cs
using System;
using System.Collections;
using System.Collections.Generic;
using DMD;
using DMD.XML;
using DMD.Databases;
using DMD.Databases.Collections;
using minidom;
using minidom.repositories;
using static minidom.Sistema;
using static minidom.Anagrafica;
using static minidom.Store;


namespace minidom
{
    public partial class Store
    {

        /// <summary>
        /// Definisce il tipo prezzo
        /// </summary>
        /// <remarks></remarks>
        public enum TipoPrezzoListino : int
        {
            /// <summary>
            /// Flag validi per le relazioni Articoli x Listino
            /// </summary>
            FISSO = 0
        }

        /// <summary>
        /// Definisce la relazione tra un articolo ed un listino prezzi
        /// </summary>
        /// <remarks></remarks>
        [Serializable]
        public class ArticoloXListino 
            : minidom.Databases.DBObject
        {
            private TipoPrezzoListino m_TipoPrezzo;
            private int m_IDArticolo;
            [NonSerialized] private Articolo m_Articolo;
            private int m_IDListino;
            [NonSerialized] private Listino m_Listino;
             
            /// <summary>
            /// Costruttore
            /// </summary>
            public ArticoloXListino()
            {
                m_IDArticolo = 0;
                m_Articolo = null;
                m_IDListino = 0;
                m_Listino = null;
                m_TipoPrezzo = TipoPrezzoListino.FISSO;
            }

            /// <summary>
            /// Restituisce o imposta il tipo prezzo
            /// </summary>
            /// <value></value>
            /// <returns></returns>
            /// <remarks></remarks>
            public TipoPrezzoListino TipoPrezzo
            {
                get
                {
                    return m_TipoPrezzo;
                }

                set
                {
                    var oldValue = m_TipoPrezzo;
                    if (oldValue == value)
                        return;
                    m_TipoPrezzo = value;
                    DoChanged("TipoPrezzo", value, oldValue);
                }
            }

            /// <summary>
            /// Restituisce o imposta l'articolo a cui è associato il codice
            /// </summary>
            /// <value></value>
            /// <returns></returns>
            /// <remarks></remarks>
            public Articolo Articolo
            {
                get
                {
                    lock (this)
                    {
                        if (m_Articolo is null)
                            m_Articolo = Articoli.GetItemById(m_IDArticolo);
                        return m_Articolo;
                    }
                }

                set
                {
                    Articolo oldValue;
                    lock (this)
                    {
                        oldValue = m_Articolo;
                        if (ReferenceEquals(oldValue, value))
                            return;
                        m_Articolo = value;
                        m_IDArticolo = DBUtils.GetID(value, 0);
                    }

                    DoChanged("Articolo", value, oldValue);
                }
            }

            /// <summary>
            /// Restituisce o imposta l'id dell'articolo
            /// </summary>
            public int IDArticolo
            {
                get
                {
                    return DBUtils.GetID(m_Articolo, m_IDArticolo);
                }

                set
                {
                    int oldValue = IDArticolo;
                    if (oldValue == value)
                        return;
                    m_IDArticolo = 0;
                    m_Articolo = null;
                    DoChanged("IDArticolo", value, oldValue);
                }
            }

            /// <summary>
            /// Imposta l'articolo
            /// </summary>
            /// <param name="value"></param>
            protected internal virtual void SetArticolo(Articolo value)
            {
                m_Articolo = value;
                m_IDArticolo = DBUtils.GetID(value, 0);
            }

            /// <summary>
            /// Restituisce o imposta il magazzino
            /// </summary>
            /// <value></value>
            /// <returns></returns>
            /// <remarks></remarks>
            public Listino Listino
            {
                get
                {
                    lock (this)
                    {
                        if (m_Listino is null)
                            m_Listino = Listini.GetItemById(m_IDListino);
                        return m_Listino;
                    }
                }

                set
                {
                    Listino oldValue;
                    lock (this)
                    {
                        oldValue = m_Listino;
                        if (ReferenceEquals(oldValue, value))
                            return;
                        m_Listino = value;
                        m_IDListino = DBUtils.GetID(value, 0);
                    }

                    DoChanged("Listino", value, oldValue);
                }
            }

            /// <summary>
            /// Restituisce o imposta l'id dell'articolo
            /// </summary>
            public int IDListino
            {
                get
                {
                    return DBUtils.GetID(m_Listino, m_IDListino);
                }

                set
                {
                    int oldValue = IDListino;
                    if (oldValue == value)
                        return;
                    m_IDListino = value;
                    m_Listino = null;
                    DoChanged("IDListino", value, oldValue);
                }
            }

             
            /// <summary>
            /// Calcola il prezzo di vendita
            /// </summary>
            /// <param name="docuemnto"></param>
            /// <returns></returns>
            public decimal CalcolaPrezzoVendita(object docuemnto)
            {
                //TODO CalcolaPrezzoVendita
                return 0m;
            }

            /// <summary>
            /// Serializzazione xml
            /// </summary>
            /// <param name="writer"></param>
            protected override void XMLSerialize(XMLWriter writer)
            {
                writer.WriteAttribute("IDArticolo", IDArticolo);
                writer.WriteAttribute("IDListino", IDListino);
                writer.WriteAttribute("TipoPrezzo", (int?)m_TipoPrezzo);
                base.XMLSerialize(writer);
            }

            /// <summary>
            /// Deserializzazione xml
            /// </summary>
            /// <param name="fieldName"></param>
            /// <param name="fieldValue"></param>
            protected override void SetFieldInternal(string fieldName, object fieldValue)
            {
                switch (fieldName ?? "")
                {
                    case "IDArticolo":
                        {
                            m_IDArticolo = (int)DMD.XML.Utils.Serializer.DeserializeInteger(fieldValue);
                            break;
                        }

                    case "IDListino":
                        {
                            m_IDListino = (int)DMD.XML.Utils.Serializer.DeserializeInteger(fieldValue);
                            break;
                        }
 
                    case "TipoPrezzo":
                        {
                            m_TipoPrezzo = (TipoPrezzoListino)DMD.XML.Utils.Serializer.DeserializeInteger(fieldValue);
                            break;
                        }

                    default:
                        {
                            base.SetFieldInternal(fieldName, fieldValue);
                            break;
                        }
                }
            }

            /// <summary>
            /// Discriminator
            /// </summary>
            /// <returns></returns>
            public override string GetTableName()
            {
                return "tbl_OfficeArticoliListino";
            }
             
            /// <summary>
            /// Repository
            /// </summary>
            /// <returns></returns>
            public override CModulesClass GetModule()
            {
                return minidom.Store.Articoli.Listini;
            }

            /// <summary>
            /// Carica dal db
            /// </summary>
            /// <param name="reader"></param>
            /// <returns></returns>
            protected override bool LoadFromRecordset(DBReader reader)
            {
                m_IDArticolo = reader.Read("IDArticolo",  m_IDArticolo);
                m_IDListino = reader.Read("IDListino",  m_IDListino);
                m_Flags = reader.Read("Flags",  m_Flags);
                m_TipoPrezzo = reader.Read("TipoPrezzo",  m_TipoPrezzo);
                return base.LoadFromRecordset(reader);
            }

            /// <summary>
            /// Salva nel db
            /// </summary>
            /// <param name="writer"></param>
            /// <returns></returns>
            protected override bool SaveToRecordset(DBWriter writer)
            {
                writer.Write("IDArticolo", IDArticolo);
                writer.Write("IDListino", IDListino);
                writer.Write("TipoPrezzo", m_TipoPrezzo);
                return base.SaveToRecordset(writer);
            }

            /// <summary>
            /// Prepara i campi
            /// </summary>
            /// <param name="table"></param>
            protected override void PrepareDBSchemaFields(DBTable table)
            {
                base.PrepareDBSchemaFields(table);

                var c = table.Fields.Ensure("IDArticolo", typeof(int), 1);
                c = table.Fields.Ensure("IDListino", typeof(int), 1);
                c = table.Fields.Ensure("TipoPrezzo", typeof(int), 1);
            }

            /// <summary>
            /// Prepara i vincoli
            /// </summary>
            /// <param name="table"></param>
            protected override void PrepareDBSchemaConstraints(DBTable table)
            {
                base.PrepareDBSchemaConstraints(table);

                var c = table.Constraints.Ensure("idxArticolo", new string[] { "IDArticolo" }, DBFieldConstraintFlags.None);
                c = table.Constraints.Ensure("idxTipo", new string[] { "IDListino", "TipoPrezzo" }, DBFieldConstraintFlags.None);
            }

            /// <summary>
            /// Restituisce una stringa che rappresenta l'oggetto
            /// </summary>
            /// <returns></returns>
            public override string ToString()
            {
                return DMD.Strings.ConcatArray(this.m_IDListino, " x ", this.m_IDArticolo);
            }

            /// <summary>
            /// Restituisce il codice hash dell'oggetto
            /// </summary>
            /// <returns></returns>
            public override int GetHashCode()
            {
                return HashCalculator.Calculate(this.m_IDArticolo);
            }

            /// <summary>
            /// Restituisce true se i due oggetti sono uguali
            /// </summary>
            /// <param name="obj"></param>
            /// <returns></returns>
            public sealed override bool Equals(Databases.DBObject obj)
            {
                return (obj is ArticoloXListino) && this.Equals((ArticoloXListino)obj);
            }

            /// <summary>
            /// Restituisce true se i due oggetti sono uguali
            /// </summary>
            /// <param name="obj"></param>
            /// <returns></returns>
            public virtual bool Equals(ArticoloXListino obj)
            {
                return base.Equals(obj)
                    && DMD.Integers.EQ((int)this.m_TipoPrezzo, (int)obj.m_TipoPrezzo)
                    && DMD.Integers.EQ(this.m_IDArticolo, obj.m_IDArticolo)
                    && DMD.Integers.EQ(this.m_IDListino, obj.m_IDListino)
                    ;
            }


        }
    }
}